File: implementations/zig/src/fen.zig
const std = @import("std");
const board = @import("board.zig");

pub const FenParser = struct {
    board_ref: *board.Board,

    pub fn init(board_ref: *board.Board) FenParser {
        return FenParser{
            .board_ref = board_ref,
        };
    }

    pub fn loadFromFen(self: *FenParser, fen_string: []const u8) !void {
        var parts = std.mem.splitScalar(u8, fen_string, ' ');

        // Parse piece placement
        const piece_data = parts.next() orelse return error.InvalidFen;
        try self.parsePiecePlacement(piece_data);

        // Parse active color
        const active_color = parts.next() orelse return error.InvalidFen;
        self.board_ref.white_to_move = std.mem.eql(u8, active_color, "w");

        // Parse castling availability
        const castling_data = parts.next() orelse return error.InvalidFen;
        try self.parseCastlingRights(castling_data);

        // Parse en passant target square
        const en_passant_data = parts.next() orelse return error.InvalidFen;
        try self.parseEnPassantTarget(en_passant_data);

        // Parse halfmove clock
        const halfmove_data = parts.next() orelse return error.InvalidFen;
        self.board_ref.halfmove_clock = std.fmt.parseInt(u16, halfmove_data, 10) catch return error.InvalidFen;

        // Parse fullmove number
        const fullmove_data = parts.next() orelse return error.InvalidFen;
        self.board_ref.fullmove_number = std.fmt.parseInt(u16, fullmove_data, 10) catch return error.InvalidFen;
    }

    fn parsePiecePlacement(self: *FenParser, piece_data: []const u8) !void {
        // Clear the board
        for (&self.board_ref.squares) |*square| {
            square.* = null;
        }

        var ranks = std.mem.splitScalar(u8, piece_data, '/');
        var rank: u8 = 7; // Start from rank 8 (index 7)

        while (ranks.next()) |rank_data| {
            var file: u8 = 0;

            for (rank_data) |char| {
                if (std.ascii.isDigit(char)) {
                    // Empty squares
                    const empty_count = char - '0';
                    file += empty_count;
                } else {
                    // Piece
                    if (file >= 8) return error.InvalidFen;

                    const piece = board.Piece.fromChar(char) orelse return error.InvalidFen;
                    const square_index = rank * 8 + file;
                    self.board_ref.squares[square_index] = piece;
                    file += 1;
                }
            }

            if (file != 8) return error.InvalidFen;
            if (rank == 0) break;
            rank -= 1;
        }
    }

    fn parseCastlingRights(self: *FenParser, castling_data: []const u8) !void {
        self.board_ref.castling_rights = board.CastlingRights{
            .white_king_side = false,
            .white_queen_side = false,
            .black_king_side = false,
            .black_queen_side = false,
        };

        if (std.mem.eql(u8, castling_data, "-")) {
            return; // No castling rights
        }

        for (castling_data) |char| {
            switch (char) {
                'K' => self.board_ref.castling_rights.white_king_side = true,
                'Q' => self.board_ref.castling_rights.white_queen_side = true,
                'k' => self.board_ref.castling_rights.black_king_side = true,
                'q' => self.board_ref.castling_rights.black_queen_side = true,
                else => return error.InvalidFen,
            }
        }
    }

    fn parseEnPassantTarget(self: *FenParser, en_passant_data: []const u8) !void {
        if (std.mem.eql(u8, en_passant_data, "-")) {
            self.board_ref.en_passant_target = null;
            return;
        }

        if (en_passant_data.len != 2) return error.InvalidFen;

        const file = en_passant_data[0] - 'a';
        const rank = en_passant_data[1] - '1';

        if (file > 7 or rank > 7) return error.InvalidFen;

        self.board_ref.en_passant_target = rank * 8 + file;
    }

    pub fn toFen(self: *FenParser, allocator: std.mem.Allocator) ![]u8 {
        var fen_parts = std.ArrayList([]const u8).empty;
        defer fen_parts.deinit(allocator);

        // Piece placement
        const piece_placement = try self.getPiecePlacement(allocator);
        try fen_parts.append(allocator, piece_placement);

        // Active color
        const active_color = if (self.board_ref.white_to_move) "w" else "b";
        try fen_parts.append(allocator, active_color);

        // Castling availability
        const castling = try self.getCastlingString(allocator);
        try fen_parts.append(allocator, castling);

        // En passant target square
        const en_passant = try self.getEnPassantString(allocator);
        try fen_parts.append(allocator, en_passant);

        // Halfmove clock
        const halfmove = try std.fmt.allocPrint(allocator, "{}", .{self.board_ref.halfmove_clock});
        try fen_parts.append(allocator, halfmove);

        // Fullmove number
        const fullmove = try std.fmt.allocPrint(allocator, "{}", .{self.board_ref.fullmove_number});
        try fen_parts.append(allocator, fullmove);

        // Join all parts with spaces
        return std.mem.join(allocator, " ", fen_parts.items);
    }

    fn getPiecePlacement(self: *FenParser, allocator: std.mem.Allocator) ![]u8 {
        var rank_strings = std.ArrayList([]const u8).empty;
        defer rank_strings.deinit(allocator);

        var rank: i8 = 7; // Start from rank 8
        while (rank >= 0) : (rank -= 1) {
            var rank_string = std.ArrayList(u8).empty;
            defer rank_string.deinit(allocator);

            var file: u8 = 0;
            while (file < 8) {
                var empty_count: u8 = 0;

                // Count consecutive empty squares
                while (file < 8 and self.board_ref.squares[@as(u8, @intCast(rank)) * 8 + file] == null) {
                    empty_count += 1;
                    file += 1;
                }

                // Add empty count to string if any
                if (empty_count > 0) {
                    try rank_string.append(allocator, '0' + empty_count);
                }

                // Add piece if present
                if (file < 8) {
                    if (self.board_ref.squares[@as(u8, @intCast(rank)) * 8 + file]) |piece| {
                        try rank_string.append(allocator, piece.toChar());
                    }
                    file += 1;
                }
            }

            try rank_strings.append(allocator, try rank_string.toOwnedSlice(allocator));
        }

        return std.mem.join(allocator, "/", rank_strings.items);
    }

    fn getCastlingString(self: *FenParser, allocator: std.mem.Allocator) ![]const u8 {
        var castling = std.ArrayList(u8).empty;
        defer castling.deinit(allocator);

        if (self.board_ref.castling_rights.white_king_side) try castling.append(allocator, 'K');
        if (self.board_ref.castling_rights.white_queen_side) try castling.append(allocator, 'Q');
        if (self.board_ref.castling_rights.black_king_side) try castling.append(allocator, 'k');
        if (self.board_ref.castling_rights.black_queen_side) try castling.append(allocator, 'q');

        if (castling.items.len == 0) {
            return try allocator.dupe(u8, "-");
        }

        return try castling.toOwnedSlice(allocator);
    }

    fn getEnPassantString(self: *FenParser, allocator: std.mem.Allocator) ![]const u8 {
        if (self.board_ref.en_passant_target) |target| {
            const file = @as(u8, 'a') + (target % 8);
            const rank = @as(u8, '1') + (target / 8);
            return try std.fmt.allocPrint(allocator, "{c}{c}", .{ file, rank });
        } else {
            return try allocator.dupe(u8, "-");
        }
    }
};

File: implementations/go/fen.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func (gs *GameState) castlingFEN() string {
	if gs.Chess960Mode {
		whiteFiles := make([]int, 0, 2)
		blackFiles := make([]int, 0, 2)
		if gs.CastlingRights[White][QueensideCastle] {
			whiteFiles = append(whiteFiles, gs.CastlingConfig.RookFile[White][QueensideCastle])
		}
		if gs.CastlingRights[White][KingsideCastle] {
			whiteFiles = append(whiteFiles, gs.CastlingConfig.RookFile[White][KingsideCastle])
		}
		if gs.CastlingRights[Black][QueensideCastle] {
			blackFiles = append(blackFiles, gs.CastlingConfig.RookFile[Black][QueensideCastle])
		}
		if gs.CastlingRights[Black][KingsideCastle] {
			blackFiles = append(blackFiles, gs.CastlingConfig.RookFile[Black][KingsideCastle])
		}

		var b strings.Builder
		for _, file := range sortedInts(whiteFiles) {
			b.WriteByte(byte('A' + file))
		}
		for _, file := range sortedInts(blackFiles) {
			b.WriteByte(byte('a' + file))
		}
		if b.Len() == 0 {
			return "-"
		}
		return b.String()
	}

	var castling strings.Builder
	if gs.CastlingRights[White][KingsideCastle] {
		castling.WriteRune('K')
	}
	if gs.CastlingRights[White][QueensideCastle] {
		castling.WriteRune('Q')
	}
	if gs.CastlingRights[Black][KingsideCastle] {
		castling.WriteRune('k')
	}
	if gs.CastlingRights[Black][QueensideCastle] {
		castling.WriteRune('q')
	}
	if castling.Len() == 0 {
		return "-"
	}
	return castling.String()
}

func sortedInts(values []int) []int {
	if len(values) == 0 {
		return values
	}
	cloned := make([]int, len(values))
	copy(cloned, values)
	for i := 0; i < len(cloned); i++ {
		for j := i + 1; j < len(cloned); j++ {
			if cloned[j] < cloned[i] {
				cloned[i], cloned[j] = cloned[j], cloned[i]
			}
		}
	}
	return cloned
}

func (gs *GameState) ToFEN() string {
	var fenParts []string

	// Board position
	var boardStr strings.Builder
	for rank := 7; rank >= 0; rank-- {
		emptyCount := 0
		for file := 0; file < 8; file++ {
			piece := gs.Board[rank][file]
			if piece.IsEmpty() {
				emptyCount++
			} else {
				if emptyCount > 0 {
					boardStr.WriteString(strconv.Itoa(emptyCount))
					emptyCount = 0
				}
				boardStr.WriteRune(piece.Symbol())
			}
		}
		if emptyCount > 0 {
			boardStr.WriteString(strconv.Itoa(emptyCount))
		}
		if rank > 0 {
			boardStr.WriteRune('/')
		}
	}
	fenParts = append(fenParts, boardStr.String())

	// Active color
	if gs.ActiveColor == White {
		fenParts = append(fenParts, "w")
	} else {
		fenParts = append(fenParts, "b")
	}

	// Castling availability
	fenParts = append(fenParts, gs.castlingFEN())

	// En passant target
	if gs.EnPassantTarget != nil {
		fenParts = append(fenParts, gs.EnPassantTarget.ToAlgebraic())
	} else {
		fenParts = append(fenParts, "-")
	}

	// Halfmove clock
	fenParts = append(fenParts, strconv.Itoa(gs.HalfmoveClock))

	// Fullmove number
	fenParts = append(fenParts, strconv.Itoa(gs.FullmoveNumber))

	return strings.Join(fenParts, " ")
}

func (gs *GameState) FromFEN(fen string) error {
	parts := strings.Fields(fen)
	if len(parts) != 6 {
		return fmt.Errorf("invalid FEN: expected 6 parts, got %d", len(parts))
	}

	// Clear the board
	for rank := 0; rank < 8; rank++ {
		for file := 0; file < 8; file++ {
			gs.Board[rank][file] = Piece{Type: Empty}
		}
	}

	// Parse board position
	ranks := strings.Split(parts[0], "/")
	if len(ranks) != 8 {
		return fmt.Errorf("invalid FEN: expected 8 ranks, got %d", len(ranks))
	}

	for rankIdx, rankStr := range ranks {
		rank := 7 - rankIdx // FEN starts from rank 8 (index 7)
		file := 0

		for _, char := range rankStr {
			if char >= '1' && char <= '8' {
				// Empty squares
				emptySquares := int(char - '0')
				file += emptySquares
			} else {
				// Piece
				piece, err := pieceFromSymbol(char)
				if err != nil {
					return fmt.Errorf("invalid piece symbol in FEN: %c", char)
				}
				if file >= 8 {
					return fmt.Errorf("invalid FEN: too many pieces in rank")
				}
				gs.Board[rank][file] = piece
				file++
			}
		}

		if file != 8 {
			return fmt.Errorf("invalid FEN: incomplete rank %d", rankIdx)
		}
	}

	// Parse active color
	switch parts[1] {
	case "w":
		gs.ActiveColor = White
	case "b":
		gs.ActiveColor = Black
	default:
		return fmt.Errorf("invalid active color in FEN: %s", parts[1])
	}

	// Parse castling rights
	gs.CastlingRights = [2][2]bool{{false, false}, {false, false}}
	gs.CastlingConfig = DefaultCastlingConfig()
	gs.Chess960Mode = false
	if file, ok := gs.FindHomeRankPiece(White, King); ok {
		gs.CastlingConfig.KingFile[White] = file
	}
	if file, ok := gs.FindHomeRankPiece(Black, King); ok {
		gs.CastlingConfig.KingFile[Black] = file
	}
	if parts[2] != "-" {
		for _, char := range parts[2] {
			switch char {
			case 'K':
				gs.CastlingRights[White][KingsideCastle] = true
			case 'Q':
				gs.CastlingRights[White][QueensideCastle] = true
			case 'k':
				gs.CastlingRights[Black][KingsideCastle] = true
			case 'q':
				gs.CastlingRights[Black][QueensideCastle] = true
			case 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H':
				rookFile := int(char - 'A')
				gs.Chess960Mode = true
				if rookFile > gs.CastlingConfig.KingFile[White] {
					gs.CastlingRights[White][KingsideCastle] = true
					gs.CastlingConfig.RookFile[White][KingsideCastle] = rookFile
				} else {
					gs.CastlingRights[White][QueensideCastle] = true
					gs.CastlingConfig.RookFile[White][QueensideCastle] = rookFile
				}
			case 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h':
				rookFile := int(char - 'a')
				gs.Chess960Mode = true
				if rookFile > gs.CastlingConfig.KingFile[Black] {
					gs.CastlingRights[Black][KingsideCastle] = true
					gs.CastlingConfig.RookFile[Black][KingsideCastle] = rookFile
				} else {
					gs.CastlingRights[Black][QueensideCastle] = true
					gs.CastlingConfig.RookFile[Black][QueensideCastle] = rookFile
				}
			default:
				return fmt.Errorf("invalid castling rights in FEN: %c", char)
			}
		}
	}

	// Parse en passant target
	if parts[3] != "-" {
		square := AlgebraicToSquare(parts[3])
		if !square.IsValid() {
			return fmt.Errorf("invalid en passant target in FEN: %s", parts[3])
		}
		gs.EnPassantTarget = &square
	} else {
		gs.EnPassantTarget = nil
	}

	// Parse halfmove clock
	halfmove, err := strconv.Atoi(parts[4])
	if err != nil {
		return fmt.Errorf("invalid halfmove clock in FEN: %s", parts[4])
	}
	gs.HalfmoveClock = halfmove

	// Parse fullmove number
	fullmove, err := strconv.Atoi(parts[5])
	if err != nil {
		return fmt.Errorf("invalid fullmove number in FEN: %s", parts[5])
	}
	gs.FullmoveNumber = fullmove
	gs.ZobristHash = computeZobristHash(gs)

	return nil
}

func pieceFromSymbol(symbol rune) (Piece, error) {
	var pieceType PieceType
	var color Color

	// Determine color (uppercase = white, lowercase = black)
	if symbol >= 'A' && symbol <= 'Z' {
		color = White
		symbol = rune(int(symbol) + 32) // Convert to lowercase
	} else if symbol >= 'a' && symbol <= 'z' {
		color = Black
	} else {
		return Piece{}, fmt.Errorf("invalid piece symbol: %c", symbol)
	}

	// Determine piece type
	switch symbol {
	case 'p':
		pieceType = Pawn
	case 'n':
		pieceType = Knight
	case 'b':
		pieceType = Bishop
	case 'r':
		pieceType = Rook
	case 'q':
		pieceType = Queen
	case 'k':
		pieceType = King
	default:
		return Piece{}, fmt.Errorf("invalid piece symbol: %c", symbol)
	}

	return NewPiece(pieceType, color), nil
}

// Standard starting position FEN
const StartingPositionFEN = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1"
